### ISSUE TITLE
Expose MPPE keys for plugin via a hook?

### ISSUE THREAD

**Issue by enaess** [paulusmack/ppp#137]
While MSCHAPv2 or EAP/MSCHAPv2 can negotiate the MPPE keys and they are needed in order to bind the higher-level with the lower level when negotiating the SSTP protocol. These keys are zero'd out by the time CCP come up, and are needed when network layer comes up in order to send the SSTP_CONNECTED message.

Do you have a suggested way of doing this, or would you like a patch to add a function to set the mppe keys, then call a mppe_keys_hook() for integration with custom plugins?

**Comment by paulusmack** [comment]
I don't have a suggested way, and providing a suitable hook seems reasonable.

**Comment by enaess** [comment]
I can get you a patch soon

**Comment by enaess** [comment]
@paulusmack I am looking through this, and the simplest would be to move the notify(auth_up_notifier, 0) on the outside of the if-block protected by the go->neg_eap || go->neg_chap || go->neg_upap in the network_phase() function, see auth.c. This would make it trigger for both local and peer completed authentication. Clients using SSTP typically specify "noauth" as the protocol itself performs mutual authentication via certificates (e.g. check of server certificate). 

If I could get a callback once the authentication has been completed, but before any network protocol packet has been sent, I believe we are golden. Hooking it in just when the MPPE keys has been set faces a challenge where the keys aren't actually valid for anything since the server hasn't seen e.g. EAPT_MSCHAPv2 SUCCESS packet. I don't want to have the sstp-plugin.so needing to track the state of ppp auth layer.

**Comment by enaess** [comment]
This code right here

```C
static void
network_phase(int unit)
{
    lcp_options *go = &lcp_gotoptions[unit];

    /* Log calling number. */
    if (*remote_number)
	notice("peer from calling number %q authorized", remote_number);

    /*
     * If the peer had to authenticate, run the auth-up script now.
     */
    if (go->neg_chap || go->neg_upap || go->neg_eap) {
	notify(auth_up_notifier, 0);
	auth_state = s_up;
	if (auth_script_state == s_down && auth_script_pid == 0) {
	    auth_script_state = s_up;
	    auth_script(_PATH_AUTHUP);
	}
    }
```

**Comment by enaess** [comment]
@paulusmack?

**Comment by paulusmack** [comment]
Given that nothing apparently actually uses auth_up_notifier yet, what you suggest seems fine.

**Comment by paulusmack** [comment]
Merged #216
